lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -28,48 +31,372 @@ namespace cudaq::opt {
+      auto gamma = cudaq::opt::factory::createFloatConstant(
+          loc, rewriter, angles.gamma, floatTy);
+      rewriter.create<quake::RzOp>(loc, gamma, ValueRange{}, arguments);
+    }
+    if (isAboveThreshold(angles.beta)) {
+      auto beta = cudaq::opt::factory::createFloatConstant(
+          loc, rewriter, angles.beta, floatTy);
+      rewriter.create<quake::RyOp>(loc, beta, ValueRange{}, arguments);
+    }
+    if (isAboveThreshold(angles.alpha)) {
+      auto alpha = cudaq::opt::factory::createFloatConstant(
+          loc, rewriter, angles.alpha, floatTy);
+      rewriter.create<quake::RzOp>(loc, alpha, ValueRange{}, arguments);
+    }
+    /// NOTE: Typically global phase can be ignored but, if this decomposition
+    /// is applied in a kernel that is called with `cudaq::control`, the global
+    /// phase will become a local phase and give a wrong result if we don't keep
+    /// track of that.
+    /// NOTE: R1-Rz pair results in a half the applied global phase angle,
+    /// hence, we need to multiply the angle by 2
+    auto globalPhase = 2.0 * phase;
+    if (isAboveThreshold(globalPhase)) {
+      auto phase = cudaq::opt::factory::createFloatConstant(
+          loc, rewriter, globalPhase, floatTy);
+      Value negPhase = rewriter.create<arith::NegFOp>(loc, phase);
+      rewriter.create<quake::R1Op>(loc, phase, ValueRange{}, arguments[0]);
+      rewriter.create<quake::RzOp>(loc, negPhase, ValueRange{}, arguments[0]);
+    }
+    rewriter.create<func::ReturnOp>(loc);
+    rewriter.restoreInsertionPoint(insPt);
+  }
+
+  OneQubitOpZYZ(const Eigen::Matrix2cd &vec) {
+    matrix = vec;
+    decompose();
+  }
+};
+
+struct KAKComponents {
+  // KAK decomposition allows to express arbitrary 2-qubit unitary (U) in the
+  // form: U = (a1 ⊗ a0) x exp(i(xXX + yYY + zZZ)) x (b1 ⊗ b0) where, a0, a1,
+  // b0, b1 are single qubit operations, and the exponential is specified by the
+  // 3 elements of canonical class vector x, y, z
+  Eigen::Matrix2cd a0;
+  Eigen::Matrix2cd a1;
+  Eigen::Matrix2cd b0;
+  Eigen::Matrix2cd b1;
+  double x;
+  double y;
+  double z;
+};
+
+const Eigen::Matrix4cd &MagicBasisMatrix() {
+  static Eigen::Matrix4cd MagicBasisMatrix;
+  MagicBasisMatrix << 1.0, 0.0, 0.0, 1i, 0.0, 1i, 1.0, 0, 0, 1i, -1.0, 0, 1.0,
+      0, 0, -1i;
+  MagicBasisMatrix = MagicBasisMatrix * M_SQRT1_2;
+  return MagicBasisMatrix;
+}
+
+const Eigen::Matrix4cd &MagicBasisMatrixAdj() {
+  static Eigen::Matrix4cd MagicBasisMatrixAdj = MagicBasisMatrix().adjoint();
+  return MagicBasisMatrixAdj;
+}
+
+const Eigen::Matrix4cd &GammaFactor() {
+  /// Gamma matrix = +1 +1 −1 +1
+  ///                +1 +1 +1 −1
+  ///                +1 −1 −1 −1
+  ///                +1 −1 +1 +1
+  static Eigen::Matrix4cd GammaT;
+  GammaT << 1, 1, 1, 1, 1, 1, -1, -1, -1, 1, -1, 1, 1, -1, -1, 1;
+  GammaT /= 4;
+  return GammaT;
+}
+
+std::tuple<Eigen::Matrix4d, Eigen::Matrix4cd, Eigen::Matrix4d>
+bidiagonalize(const Eigen::Matrix4cd &matrix) {
```

Comment:
nit: maybe worth adding some comments about what 'bidiagonalize' means and the return tuple.

Reply:
Addressed in 5ea69709ab689d05b8ef2607ac2726ae47e38474.